tianshou/policy/imitation/gail.py

```diff
@@ -0,0 +1,136 @@
+        depends on the size of available memory and the memory cost of the model;
+        should be as large as possible within the memory constraint. Default to 256.
+    :param bool action_scaling: whether to map actions from range [-1, 1] to range
+        [action_spaces.low, action_spaces.high]. Default to True.
+    :param str action_bound_method: method to bound action to range [-1, 1], can be
+        either "clip" (for simply clipping the action), "tanh" (for applying tanh
+        squashing) for now, or empty string for no bounding. Default to "clip".
+    :param Optional[gym.Space] action_space: env's action space, mandatory if you want
+        to use option "action_scaling" or "action_bound_method". Default to None.
+    :param lr_scheduler: a learning rate scheduler that adjusts the learning rate in
+        optimizer in each policy.update(). Default to None (no lr_scheduler).
+    :param bool deterministic_eval: whether to use deterministic action instead of
+        stochastic action sampled by the policy. Default to False.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.PPOPolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        actor: torch.nn.Module,
+        critic: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        dist_fn: Type[torch.distributions.Distribution],
+        expert_buffer: ReplayBuffer,
+        disc: torch.nn.Module,
+        disc_optim: torch.optim.Optimizer,
+        disc_repeat: int = 2,
+        eps_clip: float = 0.2,
+        dual_clip: Optional[float] = None,
+        value_clip: bool = False,
+        advantage_normalization: bool = True,
+        recompute_advantage: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(
+            actor, critic, optim, dist_fn, eps_clip, dual_clip, value_clip,
+            advantage_normalization, recompute_advantage, **kwargs
+        )
+        self.disc = disc
+        self.disc_optim = disc_optim
+        self.disc_repeat = disc_repeat
+        self.expert_buffer = expert_buffer
+        self.action_dim = actor.output_dim
+
+    def process_fn(
+        self, batch: Batch, buffer: ReplayBuffer, indices: np.ndarray
+    ) -> Batch:
+        """Pre-process the data from the provided replay buffer.
+
+        Used in :meth:`update`. Check out :ref:`process_fn` for more information.
+        """
+        # update reward
+        with torch.no_grad():
+            batch.rew = to_numpy(
+                -F.logsigmoid(-self.disc(batch.obs, batch.act)).flatten()
+            )
+        return super().process_fn(batch, buffer, indices)
+
+    def learn(  # type: ignore
+        self, batch: Batch, batch_size: int, repeat: int, **kwargs: Any
+    ) -> Dict[str, List[float]]:
+        # update discriminator
+        losses = []
+        acc_pis = []
+        acc_exps = []
+        for b in batch.split(len(batch) // self.disc_repeat, merge_last=True):
```

Comment:
is `//` correct?

Reply:
Good question. I started by doing the following:

```python
for _ in range(self.disc_repeat):
  for b in batch.split(batch_size, merge_last=True):
    ...
```

However, the results were terrible because the discriminator net quickly converged to almost 100% accuracy and near-zero loss, giving almost no reward signal to guide the PPO learning.

So I checked other reference implementations and "slowed down" the learning of the discriminator by reducing the number of updates, all of the sudden, the results improved dramatically. I tried two variants:

- In this version, I sample a few (between 1-4) mini batches (size: 64) from the big batch (size: 2048) and update the discriminator net:
```python
for i, b in enumerate(batch.split(batch_size, merge_last=True):
  if i > self.disc_repeat:
    break
  ...
```

- In this version, I use all the data from the big batch (size: 2048) but control the number of updates on the discriminator net:
```python
for b in batch.split(len(batch) // self.disc_repeat, merge_last=True):
  ...
```

From my experiments, the second version worked much better. So I kept it here.

Reference: In https://github.com/Khrylx/PyTorch-RL/blob/master/gail/gail_gym.py#L121-L130, it takes the whole big batch to update the discriminator just once, then go on to slice the big batch into mini batches and do PPO updates.

Reply:
I'm a little bit confused about the variable name, maybe it should change to something like `disc_update_num` or `disc_num_update_per_iter`?

Reply:
You are right. I have changed the name of the parameter to `disc_update_num`.